### PR TITLE
Making changes in InternalValueCount and InternalAvg to support Scripted metric in reducing agggregation

### PR DESCRIPTION
**Pull request by tandonks**
### Description
[Describe what this change achieves]

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Introduce system generated ingest pipeline ([#17817](https://github.com/opensearch-project/OpenSearch/pull/17817)))
 - Support create mode in pull-based ingestion and add retries for transient failures ([#18250](https://github.com/opensearch-project/OpenSearch/pull/18250)))
 - Decouple the init of Crypto Plugin and KeyProvider in CryptoRegistry ([18270](https://github.com/opensearch-project/OpenSearch/pull18270)))
+- Supporting Scripted Metric Aggregation when reducing aggregations in InternalValueCount and InternalAvg ([18288](https://github.com/opensearch-project/OpenSearch/pull18270)))
```
Comment (backslasht) [inline CHANGELOG.md:25]:
```suggestion
- Supporting Scripted Metric Aggregation when reducing aggregations in InternalValueCount and InternalAvg ([18288](https://github.com/opensearch-project/OpenSearch/pull/18288)))
```

File: server/src/main/java/org/opensearch/search/aggregations/metrics/InternalAvg.java [hunk 2]
```diff
@@ -109,9 +109,31 @@ public InternalAvg reduce(List<InternalAggregation> aggregations, ReduceContext
...
+                // If using InternalScriptedMetric in place of InternalAvg
+                Object value = ((InternalScriptedMetric) aggregation).aggregation();
+                if (value instanceof double[]) {
+                    double[] values = (double[]) value;
+                    if (values.length != 2) {
+                        throw new IllegalArgumentException(
+                            "Expected double array with [sum, count] from ScriptedMetric but got array of length " + values.length
+                        );
+                    }
+                    count += (long) values[1];  // count is at index 1
```
Comment (backslasht) [inline server/src/main/java/org/opensearch/search/aggregations/metrics/InternalAvg.java:122]:
```suggestion
                    count += Double.valueOf(values[1]).longValue();  // count is at index 1
```

File: server/src/main/java/org/opensearch/search/aggregations/metrics/InternalAvg.java [hunk 3]
```diff
@@ -109,9 +109,31 @@ public InternalAvg reduce(List<InternalAggregation> aggregations, ReduceContext
...
+                    kahanSummation.add(values[0]);  // sum is at index 0
+                } else {
+                    throw new IllegalArgumentException(
+                        "Expected double array[sum, count] from ScriptedMetric but got ["
+                            + (value == null ? "null" : value.getClass().getName())
+                            + "]"
```
Comment (backslasht) [inline server/src/main/java/org/opensearch/search/aggregations/metrics/InternalAvg.java:128]: is this user visible error string?

File: server/src/main/java/org/opensearch/search/aggregations/metrics/InternalAvg.java [hunk 4]
```diff
@@ -109,9 +109,31 @@ public InternalAvg reduce(List<InternalAggregation> aggregations, ReduceContext
...
-            InternalAvg avg = (InternalAvg) aggregation;
-            count += avg.count;
-            kahanSummation.add(avg.sum);
+            if (aggregation instanceof InternalScriptedMetric) {
+                // If using InternalScriptedMetric in place of InternalAvg
+                Object value = ((InternalScriptedMetric) aggregation).aggregation();
+                if (value instanceof double[]) {
+                    double[] values = (double[]) value;
+                    if (values.length != 2) {
+                        throw new IllegalArgumentException(
```
Comment (backslasht) [inline server/src/main/java/org/opensearch/search/aggregations/metrics/InternalAvg.java:118]: is this user visible error string?

File: server/src/test/java/org/opensearch/search/aggregations/metrics/InternalAvgTests.java [hunk 5]
```diff
@@ -113,6 +116,32 @@ protected void assertFromXContent(InternalAvg avg, ParsedAggregation parsedAggre
         }
     }
 
+    public void testReduceWithScriptedMetric() {
```
Comment (backslasht) [inline server/src/test/java/org/opensearch/search/aggregations/metrics/InternalAvgTests.java:119]: can we add test cases for negative cases as well? 
1. when the values is not double[]
2. when the value is not InternalScriptedMetric

**Review by backslasht — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/search/aggregations/metrics/InternalAvg.java (hunk 4 above)
Comment (tandonks) [reply server/src/main/java/org/opensearch/search/aggregations/metrics/InternalAvg.java:118]: Yup from what I checked, all these would be user visible. I had included them for faster debugging in the future, but do we not want to expose these logs to the user?

**Comment by tandonks** [comment]
Refer PR: https://github.com/opensearch-project/OpenSearch/pull/18411 The commit history got corrupted in this one
